https://bugs.debian.org/cgi-bin/bugreport.cgi?att=0;bug=468114;msg=32

On 2015-08-11, Vagrant Cascadian wrote:
> On 2015-08-11, Vagrant Cascadian wrote:
>> I've taken a quick stab as refreshing this, though I haven't yet tested
>> it. I did move mounting of the loopback to mount_loop_root in functions,
>> as I would like to eventually support loopback mounted files from NFS.
>
> Made some small changes, updated and *tested* this time! Patch below...

And added NFS support! Full patch below...

live well,
  vagrant

Refreshed for Ubuntu initramfs-tools 0.151ubuntu1, which already parses the
loop=, loopflags= and loopfstype= kernel command line arguments and mounts the
loop file inline in scripts/local.  What is still needed here is the exported
LOOP* variables, the mount_loop_root() helper used by the NFS path, the NFS
hook itself and the ntfs-aware read-only condition.

---
 init              |  3 +++
 scripts/functions | 40 ++++++++++++++++++++++++++++++++++++++++
 scripts/local     |  2 +-
 scripts/nfs       |  4 ++++
 4 files changed, 48 insertions(+), 1 deletion(-)

diff --git a/init b/init
--- a/init
+++ b/init
@@ -60,6 +60,9 @@
 export ROOTDELAY=
 export ROOTFLAGS=
 export ROOTFSTYPE=
+export LOOP=
+export LOOPFLAGS=
+export LOOPFSTYPE=
 export IP=
 export IP6=
 export VLAN=
diff --git a/scripts/functions b/scripts/functions
--- a/scripts/functions
+++ b/scripts/functions
@@ -1043,6 +1043,46 @@
 	echo "$result"
 }
 
+# Mount a loopback device, which is present on the mounted filesystem.
+mount_loop_root()
+{
+	mkdir -p /host
+	mount -o move "${rootmnt}" /host
+	loopfile="/host/${LOOP#/}"
+
+	while [ ! -e "$loopfile" ]; do
+		panic "ALERT! $loopfile does not exist.  Dropping to a shell!"
+	done
+
+	if [ "${readonly?}" = "y" ]; then
+		roflag=-r
+	else
+		roflag=-w
+	fi
+
+	# Get the loop filesystem type if not set
+	if [ -z "${LOOPFSTYPE}" ]; then
+	    FSTYPE=$(get_fstype "$loopfile")
+	else
+	    FSTYPE=${LOOPFSTYPE}
+	fi
+
+	# FIXME This has no error checking
+	modprobe loop
+	modprobe "${FSTYPE}"
+
+	# FIXME This has no error checking
+	if [ -z "${LOOPFLAGS}" ]; then
+		mount ${roflag} -o loop -t "${FSTYPE}" "$loopfile" "${rootmnt}"
+	else
+		mount ${roflag} -o loop -t "${FSTYPE}" "${LOOPFLAGS}" "$loopfile" "${rootmnt}"
+	fi
+
+	if [ -d "${rootmnt}/host" ]; then
+		mount -o move /host "${rootmnt}/host"
+	fi
+}
+
 # Mount the root file system.  It should be overridden by all
 # boot scripts.
 mountroot()
diff --git a/scripts/local b/scripts/local
--- a/scripts/local
+++ b/scripts/local
@@ -206,7 +206,7 @@
 	local_premount
 
 	if [ "${readonly?}" = "y" ] && \
-	   [ -z "$LOOP" ]; then
+	   { [ -z "$LOOP" ] || [ "${FSTYPE#ntfs}" = "$FSTYPE" ]; }; then
 		roflag=-r
 	else
 		roflag=-w
diff --git a/scripts/nfs b/scripts/nfs
--- a/scripts/nfs
+++ b/scripts/nfs
@@ -74,6 +74,10 @@
 
 	# shellcheck disable=SC2086
 	nfsmount -o nolock ${roflag} ${NFSOPTS} "${NFSROOT}" "${rootmnt?}"
+
+	if [ "${LOOP}" ]; then
+		mount_loop_root
+	fi
 }
 
 # NFS root mounting
-- 
2.17.1

